pkg/typesystem/typesystem_create_test.go

```diff
@@ -0,0 +1,323 @@
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	cuA := a.GetComputedUserset()
+	cuB := b.GetComputedUserset()
+
+	if cuA == nil || cuB == nil {
+		require.Equal(t, cuA, cuB)
+		return
+	}
+	objectRelationEquals(t, cuA, cuB)
+}
+
+func objectRelationEquals(t *testing.T, a, b *openfgav1.ObjectRelation) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	require.Equal(t, a.GetRelation(), b.GetRelation())
+	require.Equal(t, a.GetObject(), b.GetObject())
+}
+
+func rewriteEquals(t *testing.T, a, b *openfgav1.Userset) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+
+	usersetA := a.GetUserset()
+	usersetB := b.GetUserset()
+
+	switch x := usersetA.(type) {
+	case *openfgav1.Userset_This:
+		_, ok := usersetB.(*openfgav1.Userset_This)
+		require.True(t, ok)
+	case *openfgav1.Userset_ComputedUserset:
+		y, ok := usersetB.(*openfgav1.Userset_ComputedUserset)
+		require.True(t, ok)
+		objectRelationEquals(t, x.ComputedUserset, y.ComputedUserset)
+	case *openfgav1.Userset_TupleToUserset:
+		y, ok := usersetB.(*openfgav1.Userset_TupleToUserset)
+		require.True(t, ok)
+		tupleToUsersetEquals(t, x.TupleToUserset, y.TupleToUserset)
+	case *openfgav1.Userset_Union:
+		y, ok := usersetB.(*openfgav1.Userset_Union)
+		require.True(t, ok)
+		usersetsEquals(t, x.Union, y.Union)
+	case *openfgav1.Userset_Intersection:
+		y, ok := usersetB.(*openfgav1.Userset_Intersection)
+		require.True(t, ok)
+		usersetsEquals(t, x.Intersection, y.Intersection)
+	case *openfgav1.Userset_Difference:
+		y, ok := usersetB.(*openfgav1.Userset_Difference)
+		require.True(t, ok)
+		if x.Difference == nil || y.Difference == nil {
+			require.Equal(t, x.Difference, y.Difference)
+			return
+		}
+		rewriteEquals(t, x.Difference.GetBase(), y.Difference.GetBase())
+		rewriteEquals(t, x.Difference.GetSubtract(), y.Difference.GetSubtract())
+	default:
+		require.Equal(t, usersetA, usersetB)
+	}
+}
+
+func relationReferenceEquals(t *testing.T, a, b *openfgav1.RelationReference) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	require.Equal(t, a.GetCondition(), b.GetCondition())
+	require.Equal(t, a.GetType(), b.GetType())
+
+	rwA := a.GetRelationOrWildcard()
+	rwB := b.GetRelationOrWildcard()
+
+	switch x := rwA.(type) {
+	case *openfgav1.RelationReference_Wildcard:
+		y, ok := rwB.(*openfgav1.RelationReference_Wildcard)
+		require.True(t, ok)
+		require.Equal(t, x.Wildcard, y.Wildcard)
+	case *openfgav1.RelationReference_Relation:
+		y, ok := rwB.(*openfgav1.RelationReference_Relation)
+		require.True(t, ok)
+		require.Equal(t, x.Relation, y.Relation)
+	default:
+		require.Equal(t, rwA, rwB)
+	}
+}
+
+func relationTypeInfoEquals(t *testing.T, a, b *openfgav1.RelationTypeInfo) {
+	if a == nil || b == nil {
+		require.Equal(t, a, b)
+		return
+	}
+	directA := a.GetDirectlyRelatedUserTypes()
+	directB := b.GetDirectlyRelatedUserTypes()
+
+	if directA == nil || directB == nil {
```

Comment:
At this point, neither directA nor directB should ever be nil, correct?  In this case, maybe better to assert for non nil.

Reply:
`GetDirectlyRelatedUserTypes` returns a slice, so it most certainly can be `nil`.

Reply:
Why do you say it should never be `nil` at this point? Are you stating that the value returned from `GetDirectlyRelatedUserTypes` should never return a `nil` value? If so, we should probably create a different test for that, because this function is testing for equality, not broken assumptions.

Reply:
Sure, we can add another test for validation.  The reason I am a bit concern is that if this function is used only for TestNewTypeSystem, this should be fine as we hard code what is the expected value.  However, if it used for other validation, we may say they are correct while semantically it does not make sense .

Reply:
Sure, we can add another test for validation.  The reason I am a bit concern is that if this function is used only for TestNewTypeSystem, this should be fine as we hard code what is the expected value.  However, if it used for other validation, we may say they are correct while semantically it does not make sense .

Reply:
Why would you use a function that tests for equality when you are testing for another specific assumption?